cpp/test/core/capped_allocator_tests.cpp

```diff
@@ -0,0 +1,402 @@
+{
+    bond::capped_allocator<std::allocator<char>, bond::shared_counter<>> alloc{ 100 };
+    const auto count = alloc.get_counter().value();
+    BOOST_CHECK_NE(count, 0u);
+
+    BOOST_CHECK_THROW(alloc.allocate(alloc.get_counter().max_value() - count + 1), std::bad_alloc);
+    BOOST_CHECK_EQUAL(alloc.get_counter().value(), count);
+}
+
+BOOST_AUTO_TEST_CASE(AllocatorComparisonTest)
+{
+    auto state = std::make_shared<int>();
+
+    bond::capped_allocator<allocator_with_state<>> a1(1024, state);
+    BOOST_CHECK_EQUAL(a1.get_allocator().state, state);
+
+    bond::capped_allocator<allocator_with_state<>> a2(1024, state);
+    BOOST_CHECK_EQUAL(a2.get_allocator().state, state);
+
+    BOOST_CHECK_NE(&a1.get_counter(), &a2.get_counter());
+    BOOST_CHECK((a1 == a2));
+
+    bond::capped_allocator<allocator_with_state<>> a3(1024, std::make_shared<int>());
+    BOOST_CHECK_NE(a1.get_allocator().state, a3.get_allocator().state);
+    BOOST_CHECK_NE(a2.get_allocator().state, a3.get_allocator().state);
+    BOOST_CHECK((a1 != a3));
+    BOOST_CHECK((a2 != a3));
+}
+
+#ifdef _MSC_VER
+
+using all_protocols = boost::mpl::list<
+    bond::SimpleBinaryReader<bond::InputBuffer>,
+    bond::CompactBinaryReader<bond::InputBuffer>,
+    bond::FastBinaryReader<bond::InputBuffer> >;
+
+BOOST_AUTO_TEST_CASE_TEMPLATE(BondStructDeserializationTest, Reader, all_protocols)
+{
+    const std::size_t max_size = (std::numeric_limits<std::uint32_t>::max)();
+    bond::capped_allocator<std::allocator<char>> alloc{ max_size };
+
+    capped_allocator_tests::Struct from{ alloc };
+
+    InitRandom(from);
+    const auto obj_mem_usage = alloc.get_counter().value();
+
+    bond::blob buffer;
+
+    // BOOST_TEST_CONTEXT("Serialization without overflow")
+    {
+        using Writer = typename bond::get_protocol_writer<Reader, bond::OutputBuffer>::type;
+
+        typename Writer::Buffer output;
+        Writer writer{ output };
+        BOOST_REQUIRE_NO_THROW(bond::Serialize(from, writer));
+
+        buffer = output.GetBuffer();
+    }
+
+    // BOOST_TEST_CONTEXT("Serialization with overflow")
+    {
+        using Writer = typename bond::get_protocol_writer<Reader, bond::OutputMemoryStream<decltype(alloc)>>::type;
+
+        decltype(alloc) new_alloc{ buffer.size() / 2 };
+        typename Writer::Buffer output{ new_alloc };
+        Writer writer{ output };
+        BOOST_CHECK_THROW(bond::Serialize(from, writer), std::bad_alloc);
+    }
+
+    // BOOST_TEST_CONTEXT("Compile-time schema deserialize without overflow")
+    {
+        Reader reader{ buffer };
+
+        decltype(alloc) new_alloc{ max_size };
+        decltype(from) to{ new_alloc };
+        BOOST_REQUIRE_NO_THROW(bond::Deserialize(reader, to));
+        BOOST_CHECK((from == to));
+    }
+
+    auto expected_exception = [](const std::exception& e)
+    {
+        return dynamic_cast<const std::bad_alloc*>(&e)
+            // Some STL containers may throw std::length_error while checking
```

Comment:
Minor: if the comment is above the return, it describes what's going on fine, and the return line isn't interrupted.

Reply:
Updated.